src/hotspot/share/jvmci/jvmciRuntime.cpp

```diff
@@ -1784,6 +1785,8 @@ JVMCI::CodeInstallResult JVMCIRuntime::register_method(JVMCIEnv* JVMCIENV,
             MutexLocker ml(CompiledMethod_lock, Mutex::_no_safepoint_check_flag);
             if (nm->make_in_use()) {
               method->set_code(method, nm);
+            } else {
+              result = JVMCI::nmethod_reclaimed;
```

Comment:
Is it really possible for the make_in_use call to fail if the nmethod is nmethodLocker'ed?  The copy of this logic in ciEnv.cpp doesn't treat a failure of this call to be a compilation failure but it certainly seems like it should.  It would be better if it weren't possible for this call to fail.

Reply:
The issue is that the nmethod is published to Method code pointers before it is made in_use. That means a concurrent JavaThread may enter it and make the nmethod not_entrant, before it is made in_use. Making it in_use after it became not_entrant would cause a non-monotonic state transition which isn't great and comes with problems.
I agree it feels like there is room for improvement here, but at least that is the reason why make_in_use can fail; to prevent non-monotonic state transitions in this race.

Reply:
Ok.  I see the intent. In that case it seems like the reordering of the in_use and set_code calls was the important part of JDK-8226705.  It can't actually fail here since it hasn't been published until the following line, right?  Or is there some other way the nmethod is published?  Why couldn't the in_use state be set at the end of the nmethod factory method?

Reply:
I think it's published in the code cache and we had trouble with code walkers using only the CodeCache_lock to shoot down nmethods, but not the Compile_lock. If the walkers used the Compile_lock too then they wouldn't pick these up until they are fully installed. Plenty of room for improvements...

Reply:
Yes, directly scanning the code cache seemed like the only other path.  It's unfortunate that the nmethod lifecycle seems to have gotten more complicated instead of less so.  So it could really fail in some exceptionally weird circumstances.

Reply:
> Yes, directly scanning the code cache seemed like the only other path.  It's unfortunate that the nmethod lifecycle seems to have gotten more complicated instead of less so.  So it could really fail in some exceptionally weird circumstances.

Yes indeed. This is what motivated my work on https://openjdk.java.net/jeps/8221828
In my prototype, after removing inline caches, replacing it with an optimized table driven approach to method invocations, I could reduce the nmethod lifecycle down to a boolean: is_in_use(), which is true or false. No zombies, no unloaded, no unloading, no is_alive, no transient installation state, no separate "nmethod locking mechanism". I hope it hits mainline one day, because it reduces so much risk in our code base, by removing a huge amount of state machinery.